executor/adapter.go

```diff
@@ -1485,10 +1485,29 @@ func (a *ExecStmt) LogSlowQuery(txnTS uint64, succ bool, hasMoreResults bool) {
 	cfg := config.GetGlobalConfig()
 	costTime := time.Since(sessVars.StartTime) + sessVars.DurationParse
 	threshold := time.Duration(atomic.LoadUint64(&cfg.Instance.SlowThreshold)) * time.Millisecond
+	inEfficientScanRatio := atomic.LoadUint64(&cfg.Instance.InefficientScanRatio)
+	inEfficientScanNum := atomic.LoadUint64(&cfg.Instance.InefficientScanNum)
 	enable := cfg.Instance.EnableSlowLog.Load()
+
+	// Check if it's an expensive mvcc scan. It's regarded as expensive if
+	// 1. The rocksdb_delete_skipped_count > processed_keys * inEfficientScanRatio.
+	// 2. The iterated key number > inEfficientScanNum.
+	execDetail := stmtCtx.GetExecDetails()
+	isInefficientScan := false
+	if execDetail.ScanDetail != nil && sessVars.ConnectionID > 0 {
+		processedKeys := execDetail.ScanDetail.ProcessedKeys
+		if processedKeys <= 0 {
+			processedKeys = 1
+		}
+		currentScanRatio := execDetail.ScanDetail.RocksdbKeySkippedCount / uint64(processedKeys)
+		if currentScanRatio >= inEfficientScanRatio && execDetail.ScanDetail.RocksdbKeySkippedCount >= inEfficientScanNum {
+			isInefficientScan = true
+			execDetail.InEfficientScanRatio = currentScanRatio
```

Comment:
Not very important in this PR though, IMO it doesn't look to be a very good idea to calculate some fields of the `ExecDetails` at this place. I think the field can be calculated when the `ExecDetails` is generated rather than when used (maybe in `MergeExecDetails()`? I'm not sure). I think we'd better consider this when we plan to port this PR to master and official release branches.

Reply:
Yes, currently the scan details are not considered in `snapshot.get` and `snapshot.batchGet`, they should also be too if it's to be port to the master branch.